components/feature/customtabs/src/main/java/mozilla/components/feature/customtabs/CustomTabWindowFeature.kt

```diff
@@ -57,16 +57,29 @@ class CustomTabWindowFeature(
     }
 
     /**
-     * Starts the feature and a observer to listen for window requests.
+     * Starts observing the configured session to listen for window requests.
      */
     override fun start() {
-        sessionManager.findSessionById(sessionId)?.register(windowObserver)
+        scope = store.flowScoped { flow ->
+            flow.mapNotNull { state -> state.findCustomTab(sessionId) }
+                .ifChanged {
+                    it.content.windowRequest
```

Comment:
`ifChanged` is using referential equality, not structural one (e.g. iiuc it's just checking that this is the same object reference, not that its contents haven't changed).

Is it possible that emitters of `UpdateWindowRequestAction` will reuse instances of `windowRequest`, and just change wrapped `url`? If that's possible, I don't think this check will behave correctly.

Reply:
That's interesting. All our migrated features right now (prompts, downloads, etc.) work like this. We're not re-using request objects and they should be immutable for that reason! See: https://github.com/mozilla-mobile/android-components/blob/master/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/window/GeckoWindowRequest.kt and https://github.com/mozilla-mobile/android-components/blob/master/components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/PromptRequest.kt

If we find some that are mutable, we should file as a bug, but I guess immutability should definitely required for any state in our new browser-store. If we don't yet point that out in the docs, we should add some exclamation marks around it :)

Reply:
For `WindowRequest`, we should be fine though and this should behave correctly.

Reply:
Thanks for checking! I agree that as long as things are immutable, this should be fine.

> If we don't yet point that out in the docs, we should add it with some exclamation marks around it :)

:+1: :scroll: absolutely, yeah.